### ISSUE TITLE
Trilinos nightly build errors with Cuda backend, Panzer 

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#5287]
**Describe the bug**
The following build errors began with Cuda builds in the Panzer package of Trilinos:
```
20:45:34 /home/jenkins/weaver-new/workspace/KokkosEco_Trilinos_Weaver_CUDA101_opt/Trilinos/kokkos/containers/src/Kokkos_DynRankView.hpp(1768): error: no instance of constructor "Kokkos::LayoutContiguous<Layout, Stride>::LayoutContiguous [with Layout=PHX::DefaultDevLayout, Stride=32U]" matches the argument list
20:45:34           detected during:
20:45:34             instantiation of "Kokkos::ViewTraits<DataType *******, Properties...>::array_layout Kokkos::DynRankView<DataType, Properties...>::layout() const [with DataType=const panzer::Traits::Tangent::ScalarT, Properties=<PHX::DefaultFadLayout, Kokkos::DefaultExecutionSpace>]" 
20:45:34 /home/jenkins/weaver-new/workspace/KokkosEco_Trilinos_Weaver_CUDA101_opt/Trilinos/packages/sacado/src/Kokkos_DynRankView_Fad.hpp(1188): here
20:45:34             instantiation of "Kokkos::DynRankView<T, P...>::HostMirror Kokkos::create_mirror(const Kokkos::DynRankView<T, P...> &, std::enable_if<<expression>, void>::type *) [with T=const panzer::Traits::Tangent::ScalarT, P=<PHX::DefaultFadLayout, Kokkos::DefaultExecutionSpace>]" 
20:45:34 (2192): here
20:45:34             instantiation of "std::enable_if_t<<expression>, Kokkos::DynRankView<T, P...>::HostMirror> Kokkos::create_mirror_view(const Kokkos::DynRankView<T, P...> &) [with T=const panzer::Traits::Tangent::ScalarT, P=<PHX::DefaultFadLayout, Kokkos::DefaultExecutionSpace>]" 
20:45:34 /home/jenkins/weaver-new/workspace/KokkosEco_Trilinos_Weaver_CUDA101_opt/Trilinos/packages/panzer/disc-fe/src/evaluators/Panzer_ReorderADValues_Evaluator_impl.hpp(215): here
20:45:34             instantiation of "void panzer::ReorderADValues_Evaluator<TRAITS::Jacobian, TRAITS>::evaluateFields(TRAITS::EvalData) [with TRAITS=panzer::Traits]" 
20:45:34 /home/jenkins/weaver-new/workspace/KokkosEco_Trilinos_Weaver_CUDA101_opt/Trilinos/packages/panzer/disc-fe/src/evaluators/Panzer_ReorderADValues_Evaluator.cpp(52): here
20:45:34 
20:45:34 [ 67%] Building CXX object packages/ifpack2/src/CMakeFiles/ifpack2.dir/Ifpack2_Details_LinearSolverFactory_DOUBLE_INT_LONG_LONG_SERIAL.cpp.o
20:45:34 1 error detected in the compilation of "/tmp/tmpxft_00006d40_00000000-6_Panzer_ReorderADValues_Evaluator.cpp1.ii".
20:45:34 make[2]: *** [packages/panzer/disc-fe/src/CMakeFiles/panzer-disc-fe.dir/evaluators/Panzer_ReorderADValues_Evaluator.cpp.o] Error 1
20:45:34 make[2]: *** Waiting for unfinished jobs....
```

One of the changes below merged 7/27/2022 resulted in some incompatibility likely with Sacado and the LayoutContiguous type layout:
Git (git https://github.com/kokkos/kokkos.git)

    Fix warning about using host function in DynRankView ([commit: 495fa94](https://github.com/kokkos/kokkos/commit/495fa943a863ca2dafbb54a7ff285ec6388bbfaf)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Fix warning about integer conversion ([commit: fcecb6a](https://github.com/kokkos/kokkos/commit/fcecb6a5ba918d420ad0ae0f1984c25721097876)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Fix warning about integral to poitner conversion ([commit: 8e02578](https://github.com/kokkos/kokkos/commit/8e02578e45986e8f8f48a2c226649450f4435512)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Fix warnings in MSVC+NVCC about preprocessor of not ([commit: 4ae672f](https://github.com/kokkos/kokkos/commit/4ae672fceb1bd5727ff8b4a4983288e71eaf8e64)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Rename VOID tag type to void_tag ([commit: b5ce7c0](https://github.com/kokkos/kokkos/commit/b5ce7c0a863c2d4450cb95dbeb442b195fd20dd5)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Remove Kokkos::Rank limit to 6 ranks ([commit: a53ec9b](https://github.com/kokkos/kokkos/commit/a53ec9bf28f61cf4260ec489e62fe7b5e458dc00)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Do not forget to set last element to nullptr when removing flag ([commit: bc7adbf](https://github.com/kokkos/kokkos/commit/bc7adbf72f2fe8642c7c1ad6deea675a6565f129)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Fix flag removal in Tools and warn when flag is not recognized ([commit: c53ca75](https://github.com/kokkos/kokkos/commit/c53ca751a03527b36580dabe5cc5bb5f7fd4e0f5)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Test flag removal ([commit: 8542fdc](https://github.com/kokkos/kokkos/commit/8542fdcd3e29f31c769be77a4a7510ed65130520)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Add test for kokkos-tools parsing --kokkos-tools-libs flag ([commit: e8c5024](https://github.com/kokkos/kokkos/commit/e8c5024e85ff21e20bc649146a6bc1bb90559d18)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    #5274 DynamicView: Properly resize mirror instances after construction ([commit: 636727d](https://github.com/kokkos/kokkos/commit/636727dd3f9024af293742cbc6d75d2b1b0425d8)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    #5274: Fix test to match expectation of fences that need to be there ([commit: 23fb091](https://github.com/kokkos/kokkos/commit/23fb091bf7e89e343c3fd54a57b2b5c8c74afaf8)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))
    Fix DynRankView missing return statement ([commit: b33fa93](https://github.com/kokkos/kokkos/commit/b33fa93874f4647610d2f4c864c2bf96faf84254)) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Weaver_CUDA101_opt/596/changes#detail))

**Please also include the following items to support reproducing the bug**
**Reproducer** (weaver rhel7W queue):
```bash
# Repos
git clone -b kokkos-promotion https://github.com/trilinos/Trilinos.git
git clone -b develop kokkos
git clone -b develop kokkos-kernels
# Symbolic link to your kokkos and kokkos-kernels repos in Trilinos source directory for source override
cd Trilinos
ln -s <path-to-your-repo>/kokkos kokkos
ln -s <path-to-your-repo>/kokkos-kernels kokkos-kernels

cd $HOME
mkdir -p build
cd build

# Modules and env
export ATDM_CONFIG_REGISTER_CUSTOM_CONFIG_DIR=$TRILINOS_DIR/cmake/std/atdm/contributed/weaver
source $TRILINOS_DIR/cmake/std/atdm/load-env.sh weaver-cuda-10.1-opt
export ATDM_CONFIG_USE_NINJA=OFF
module list
export OMPI_CXX=<path-to-your-kokkos-repo>/bin/nvcc_wrapper

# Configure
cmake \
 -G"Unix Makefiles" \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DCMAKE_INSTALL_PREFIX=$PWD/install \
 -DCMAKE_CXX_STANDARD="14" \
 -DTrilinos_ENABLE_TESTS=OFF \
 -DTrilinos_ENABLE_ALL_PACKAGES=ON \
 -DTrilinos_ENABLE_Panzer=ON \
 -DPanzer_ENABLE_TESTS=ON \
 -DKokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
 -DKokkosKernels_SOURCE_DIR_OVERRIDE:STRING=kokkos-kernels \
$TRILINOS_DIR
```

**Comment by ndellingwood** [comment]
@ajpowelsnl @dalg24 @crtrott I marked as Blocks Promotion in case the release candidate branch was created after merge of the commits listed in OP

**Comment by masterleinad** [comment]
@ndellingwood I wasn't able to reproduce this. In case you still see it, can you try if
```diff
diff --git a/containers/src/Kokkos_DynRankView.hpp b/containers/src/Kokkos_DynRankView.hpp
index 42e1e11e2..65b9b57d8 100644
--- a/containers/src/Kokkos_DynRankView.hpp
+++ b/containers/src/Kokkos_DynRankView.hpp
@@ -1765,7 +1765,8 @@ KOKKOS_INLINE_FUNCTION constexpr auto DynRankView<D, P...>::layout() const ->
           Kokkos::abort(
               "Calling DynRankView::layout on DRV of unexpected rank");)
   }
-  return typename traits::array_layout{};
+  // control flow should never reach here
+  return m_map.layout();
 }
 
 /** \brief  Deep copy a value from Host memory into a view.  */
```
helps?

**Comment by ndellingwood** [comment]
@masterleinad thanks for the suggestion. The error occurs with Power9+Volta70 with cuda/10.1.105, build in progress with your suggestion (it's a slow build)

**Comment by ndellingwood** [comment]
@masterleinad the Cuda build is looking good (Panzer complete, 91% through without error) with your suggested change. Would you like to put in a PR to release-candidate-3.7.00 or would you like me to submit?

**Comment by masterleinad** [comment]
@ndellingwood I opened #5308 and #5309 to fix this.

**Comment by PhilMiller** [comment]
Cherry-picked fix was released in 3.7.00
